### ISSUE TITLE
Implement a locking solution for ledger channels

### ISSUE THREAD

**Issue by lalexgap** [statechannels/go-nitro#323]
Once we have a failing test from #322 we should implement a locking solution for ledger channels that addresses that.

**Comment by lalexgap** [comment]
This is dependent on #374 . Without any concurrency there is no need for locking.

**Comment by lalexgap** [comment]
Closing in favour of #374. It's simpler to track everything under #374 and split it into separate issues when we're ready to start.

**Comment by geoknee** [comment]
I looked into this again today.

The motivation now is that:
1. our engine is currently single-threaded,
2. this prevents race conditions with reading/writing objectives, channels and consensus channels
3. it is difficult for us to refactor the engine run loop without risking those race conditions popping up,

Having a clear set of guards (i.e. mutexes) in the store should allow for any concurrency pattern in the engine. 


#374 is a good start towards what we would need. But I think the most challenging thing to support is to have a consensus channel locked for the duration of a (read,update,crank,write) cycle. This requires some coordination between the store getters and setters as well as the mutexes. 

In the past we have used DBs with a `GetForUpdate(id,callback)` API, which could work well here. Adding a method to the store like that means the store can ensure the correct resources are locked for the duration of the callback. 

This task is made more complex by our two-stage read/write process (when we get/set an objective, we have to get/set all of the channels it references).
